notebook/services/contents/fileio.py

```diff
@@ -249,11 +259,24 @@ def _read_notebook(self, os_path, as_version=4):
             try:
                 return nbformat.read(f, as_version=as_version)
             except Exception as e:
+                e_orig = e
+
+            # If use_atomic_writing is enabled, we'll guess that it was also
+            # enabled when this notebook was written and look for a valid
+            # atomic intermediate.
+            tmp_path = path_to_intermediate(os_path)
+
+            if not self.use_atomic_writing or not os.path.exists(tmp_path):
                 raise HTTPError(
                     400,
-                    u"Unreadable Notebook: %s %r" % (os_path, e),
+                    u"Unreadable Notebook: %s %r" % (os_path, e_orig),
                 )
 
+            # Move the bad file aside, restore the intermediate, and try again.
+            copy2_safe(os_path, path_to_invalid(os_path))
```

Comment:
Yeah, we decided that's 'atomic enough'.

Our `atomic_writing` context manager is a bit of a misnomer now. We initially had a more truly atomic writing mechanism, but it caused problems with things like Dropbox and network folders, so we switched to something simpler but not as robust.